### PR TITLE
wgpu-core implementation for external textures

### PR DESCRIPTION
**Pull request by jamienicol**
**Connections**
Depends on #7822 
Part of #4386

**Description**
Add wgpu and wgpu-core implementation for external texture. wgpu-hal and naga implementations to follow, so for now this isn't particularly useful. But it does work with the noop backend, as can be seen in the added validation tests

**Testing**
Noop backend validation tests. Local testing. CTS testing can follow when more pieces are in place.

**Squash or Rebase?**

Rebase. If there are `fixup!` commits after review then squash them into the appropriate commit first

**Checklist**

- [x] Run `cargo fmt`.
- [x] Run `taplo format`.
- [x] Run `cargo clippy --tests`. If applicable, add:
  - [x] `--target wasm32-unknown-unknown`
- [x] Run `cargo xtask test` to run tests.
- [ ] If this contains user-facing changes, add a `CHANGELOG.md` entry. 

Don't think this is changelog worthy yet

### CODE REVIEW

File: wgpu-core/src/device/resource.rs [hunk 1]
```diff
@@ -173,6 +173,10 @@ pub struct Device {
     // Optional so that we can late-initialize this after the queue is created.
     pub(crate) timestamp_normalizer:
         OnceCellOrLock<crate::timestamp_normalization::TimestampNormalizer>,
+    /// Uniform buffer containing [`ExternalTextureParams`] with values such
+    /// that a [`TextureView`] bound to a [`wgt::BindingType::ExternalTexture`]
+    /// binding point will be rendered correctly.
+    pub(crate) default_external_texture_params_buffer: std::sync::OnceLock<Arc<Buffer>>,
```
Comment (teoxoy) [inline wgpu-core/src/device/resource.rs]: Unfortunately the device can't hold wgpu-core buffers since buffers hold a strong ref to devices. This should be a hal buffer, you can use `zero_buffer` as inspiration.

**Review by teoxoy — CHANGES_REQUESTED** [review]
Looks good overall!

File: wgpu-core/src/device/resource.rs (hunk 1 above)
Comment (jamienicol) [reply wgpu-core/src/device/resource.rs]: Makes sense, good catch. Fixed in e2375e4845d44c58ad162c197f1e1fa8fef3759c

Comment (teoxoy) [reply wgpu-core/src/device/resource.rs]: Fixup commit LGTM!

**Comment by teoxoy** [comment]
This PR should be good to go but we should wait for https://github.com/gfx-rs/wgpu/pull/7822 to land first.

**Review by teoxoy — APPROVED** [review]

**Comment by jamienicol** [comment]
In addition to waiting for https://github.com/gfx-rs/wgpu/pull/7822 to land first, let me rebase this manually too before merging. (I'm pretty sure my fixup commit needs split and squashed into two different ones in the series)

**Comment by jamienicol** [comment]
I've rebased this, and squashed the fixup commit properly.

When adding tests for a PR which will follow, I noticed a slight issue. The `sample_transform` member of `ExternalTextureParams` is used to transform sample coordinates, allowing for things like crop rects, rotation, etc. I originally planned to use this for both samples and loads (doing some extra maths in the shader for loads). But my tests found a corner case which the CTS didn't, and I now realise we need a separate transform for loads. Less maths in the shader should also be better performance wise too. So I've updated `ExternalTextureParams` and `ExternalTextureDescriptor` to include that.

I'll also update #7826 accordingly

**Comment by jamienicol** [comment]
Sorry for the noise, did not mean to link those issues!

Rebased patches, needed fixes to implement for WebGPUErrorType for CreateExternalTextureError, and for some changes with Hal::BufferBinding API.

also added functions for destroying external textures, and creating external textures in error state

**Comment by jimblandy** [comment]
Since I've landed #7822, I rebased this on trunk.
